sdk/tools/azure-openrewrite/src/main/java/com/azure/openrewrite/core/http/HttpHeadersCustomRecipe.java

```diff
@@ -0,0 +1,198 @@
+import org.openrewrite.Recipe;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.JavaTemplate;
+import org.openrewrite.java.MethodMatcher;
+import org.openrewrite.java.tree.J;
+
+import com.azure.openrewrite.util.ConfiguredParserJavaTemplateBuilder;
+
+/**
+ * A custom OpenRewrite recipe to migrate the use of HttpHeaders.
+ *
+ * <p>This recipe performs the following transformations:</p>
+ * <ul>
+ *   <li>Replaces the add method of HttpHeaders to use HttpHeaderName instead of String.</li>
+ *   <li>Replaces the set method of HttpHeaders to use HttpHeaderName instead of String.</li>
+ * </ul>
+ *
+ * <p>Example transformations:</p>
+ * <pre>
+ * Before: com.azure.core.http.HttpHeaders add(java.lang.String, java.lang.String)
+ * After: io.clientcore.core.http.models.HttpHeaders add(io.clientcore.core.http.models.HttpHeaderName, java.lang.String)
+ *
+ * Before: com.azure.core.http.HttpHeaders set(java.lang.String, java.lang.String)
+ * After: io.clientcore.core.http.models.HttpHeaders set(io.clientcore.core.http.models.HttpHeaderName, java.lang.String)
+ * </pre>
+ */
+public class HttpHeadersCustomRecipe extends Recipe {
+
+    /**
+     * Default constructor for {@link HttpHeadersCustomRecipe}.
+     */
+    public HttpHeadersCustomRecipe() {
+        super();
+    }
+    @Override
+    public String getDisplayName() {
+        return "HttpRequestHeaderCustomRecipe";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Convert String Literals to HttpHeaders.";
+    }
+
+    @Override
+    public JavaIsoVisitor<ExecutionContext> getVisitor() {
+        ConfiguredParserJavaTemplateBuilder templateBuilder = ConfiguredParserJavaTemplateBuilder.defaultBuilder();
+        return new JavaIsoVisitor<ExecutionContext>() {
+
+            // replace core HttpHeader methods calls with clientcore HttpHeader method calls
+            @Override
+            public J.MethodInvocation visitMethodInvocation(J.MethodInvocation method, ExecutionContext executionContext) {
+                method = super.visitMethodInvocation(method, executionContext);
+
+                MethodMatcher methodMatcher;
+                JavaTemplate replacementTemplate;
+
+                // before: com.azure.core.http.HttpHeaders HttpHeaders int getSize()
+                // after: no change
+
+                // before: com.azure.core.http.HttpHeaders HttpHeaders add(java.lang.String, java.lang.String)
+                // after: io.clientcore.core.http.models.HttpHeaders HttpHeaders add(io.clientcore.core.http.models.HttpHeaderName, java.lang.String)
+                methodMatcher = new MethodMatcher("com.azure.core.http.HttpHeaders add(java.lang.String, java.lang.String)");
+                if (methodMatcher.matches(method, false)) {
+                    replacementTemplate = templateBuilder.getJavaTemplateBuilder("add(#{any(io.clientcore.core.http.models.HttpHeaderName)}, #{any(java.lang.String)})")
```

Comment:
Can this template be build once and reused?

Reply:
@samvaity No because the `code` portion of the template is different on each use. The parser within the template however may be able to be static.